### PR TITLE
(RE-5646) add_source unpacks compressed second source & added spec tests

### PR DESCRIPTION
**Pull request by sarameisburger**
Before, if a second source was compressed it could not be unpacked because we assumed it was along the lines of a conf file. Now handles second sources that are archives and unpacks them appropriately. Spec tests were also added to ensure that the secondary source was added and unpacked properly.

Previous discussion of this PR can be found [here](https://github.com/puppetlabs/vanagon/pull/448)

### CODE REVIEW

**Comment by mckern** [comment]
Rebase & squash before review?

**Comment by underscorgan** [comment]
@sarameisburger it looks like in the squashing you picked up all my garbage commit messages, we should clean those up :O

File: lib/vanagon/component.rb [hunk 1]
```diff
@@ -231,13 +231,18 @@ def get_build_dir
     # Fetches secondary sources for the component. These are just dumped into the workdir currently.
     #
     # @param workdir [String] working directory to put the source into
-    def get_sources(workdir)
+    def get_sources(workdir) # rubocop:disable Metrics/AbcSize
+      # create an empty array if @extract_with doesn't already exist
+      @extract_with = Array(@extract_with)
```
Comment (mckern) [inline lib/vanagon/component.rb]: Using `[@extract_with]` would be a bit more idiomatic than `Array(@extract_with)`. Not a deal breaker, but worth pointing out. Also, use the accessor (`the #extract_with` method) instead of the instance variable (`@extract_with`).

File: lib/vanagon/component.rb [hunk 2]
```diff
@@ -231,13 +231,18 @@ def get_build_dir
...
       sources.each do |source|
         src = Vanagon::Component::Source.source(
           source.url, workdir: workdir, ref: source.ref, sum: source.sum
         )
         src.fetch
         src.verify
+        # set src.file to only be populated with the basename instead of entire file path
+        src.file = File.basename(src.url)
+        @extract_with << src.extract(platform.tar) if src.respond_to?(:extract)
```
Comment (mckern) [inline lib/vanagon/component.rb]: Same as above: accessor, not instance variable.

File: spec/lib/vanagon/component_spec.rb [hunk 3]
```diff
@@ -45,20 +46,32 @@
...
 
     it "copies secondary sources into the workdir" do
-      subject.get_sources(@workdir)
+      component = subject
+      component.get_sources(@workdir)
       expect(File.exist?(File.join(@workdir, @file_name))).to be true
+      # make sure that our secondary source(s) made it into the workdir
+      expect(File.exist?(File.join(@workdir, "#{@fake_dir}.tar.gz"))).to be true
+      expect(component.extract_with.join(" && ")).to match "#{@fake_dir}.tar.gz"
     end
```
Comment (mckern) [inline spec/lib/vanagon/component_spec.rb]: Instead of modifying the existing test, considering adding a new test that validates that the compressed files are expanded. Leave the existing test for uncompressed files alone, because the expectation is that secondary sources can be either kind of file.

**Comment by mckern** [comment]
Why the new PR?

File: lib/vanagon/component.rb (hunk 1 above)
Comment (underscorgan) [inline lib/vanagon/component.rb]: May not be an issue since we should only get_sources once, but was mostly using `Array` because `foo = ['a']; Array(foo) = ['a']` whereas `[foo] = [['a']]`

File: spec/lib/vanagon/component_spec.rb (hunk 3 above)
Comment (underscorgan) [inline spec/lib/vanagon/component_spec.rb]: Ok, that could be moved into a new test, though the existing test would need some extra stubbing at least so it will properly have a platform. By 'validates the compressed files are expanded' do you mean as it is tested here, or test that the makefile rules run and the tarball is extracted?

**Comment by sarameisburger** [comment]
@mckern there was a branch issue so I thought it would be safer to just make a new PR. I'll start on those other changes.

File: lib/vanagon/component.rb (hunk 1 above)
Comment (mckern) [inline lib/vanagon/component.rb]: I didn't realize you were coercing the value of `@extract_with` to an empty array, even if it's already an array. Wouldn't that be better handled by using the accessor than the instance variable, and using the accessor to ensure that `extract_with` returns what you need it to?

```ruby
# example only; didn't do the research
def extract_with
  @extract_with ||= []
end
```

**Comment by mckern** [comment]
Branch issue? Like a merge conflict?

**Comment by mckern** [comment]
As per our conversation on Monday, your new accessor could be simplified a bit by taking advantage of Ruby's implicit returns and automatic accessors:

```ruby
# This is WAY cut down from the normal class, but it should (hopefully)
# illustrate some of the stuff I talked about on Monday
class Vanagon
  class Component
    attr_writer :extract_with

    def get_source(workdir)
      if url
        # <---✂---✂ snip ✂---✂--->
        if source.respond_to?(:extract)
          extract_with << source.extract(platform.tar)
        end
      else
        # <---✂---✂ snip ✂---✂--->
        extract_with << ': no source, so nothing to extract'
      end
    end

     def get_sources(workdir) # rubocop:disable Metrics/AbcSize
      # <---✂---✂ snip ✂---✂--->
      extract_with << src.extract(platform.tar) if src.respond_to?(:extract)
    end
 
    # @return [Array]
    def extract_with
      @extract_with ||= []
    end
  end
```

File: spec/lib/vanagon/component_spec.rb (hunk 3 above)
Comment (mckern) [inline spec/lib/vanagon/component_spec.rb]: @underscorgan I may dig into this test today so we can tie this PR off

**Comment by underscorgan** [comment]
I want to just run a quick test for this.

**Comment by mckern** [comment]
No rush.

**Comment by underscorgan** [comment]
Ok, rendering off of a hacky branch that adds some files in a tar as a secondary results in:
```
pe-activemq-unpack: file-list-before-build
  gunzip -c "apache-activemq-5.14.3-bin.tar.gz" | tar xf - && \
  : nothing to extract && \
  : nothing to extract && \
  gunzip -c "pe-activemq-properties.tar.gz" | tar xf - && \
  : nothing to extract && \
  : nothing to extract && \
  : nothing to extract && \
  : nothing to extract
  touch pe-activemq-unpack
```

which is what I expect. HOORAY 🎉

File: lib/vanagon/component.rb [hunk 4]
```diff
@@ -29,7 +29,8 @@ class Component
     # that they were cloned to. For the outlying flat files, it'll
     # end up being defined explicitly as the string './'
     attr_accessor :dirname
-    # what special tool should be used to extract the primary source
+    # Which specific tool or command line invocations that
+    # should be used to extract a given component's primary source?
```
Comment (mwaggett) [inline lib/vanagon/component.rb]: I don't understand this comment, mostly bc it seems ungrammatical.

File: lib/vanagon/component.rb [hunk 5]
```diff
@@ -231,16 +233,25 @@ def get_build_dir
     # Fetches secondary sources for the component. These are just dumped into the workdir currently.
     #
     # @param workdir [String] working directory to put the source into
-    def get_sources(workdir)
+    def get_sources(workdir) # rubocop:disable Metrics/AbcSize
```
Comment (mwaggett) [inline lib/vanagon/component.rb]: probably don't want this comment here?

Comment (melissa) [inline lib/vanagon/component.rb]: We need this comment to tell rubocop that's it's okay this method is failing checks and to not error out because of it

File: lib/vanagon/component.rb [hunk 6]
```diff
@@ -231,16 +233,25 @@ def get_build_dir
...
       sources.each do |source|
         src = Vanagon::Component::Source.source(
           source.url, workdir: workdir, ref: source.ref, sum: source.sum
         )
         src.fetch
         src.verify
+        # set src.file to only be populated with the basename instead of entire file path
+        src.file = File.basename(src.url)
+        extract_with << src.extract(platform.tar) if src.respond_to? :extract
       end
```
Comment (mwaggett) [inline lib/vanagon/component.rb]: I'm confused as to why we have get_source AND get_sources. A lot of the behavior looks to be the same, so there could probably be some DRYing up of this.

Comment (underscorgan) [inline lib/vanagon/component.rb]: Yes, I agree, though tend to think that will be a bigger change and probably outside the scope of this PR.

File: lib/vanagon/component.rb (hunk 5 above)
Comment (mwaggett) [inline lib/vanagon/component.rb]: oookay

File: lib/vanagon/component.rb (hunk 6 above)
Comment (mwaggett) [inline lib/vanagon/component.rb]: It just seems like one of those things where if we don't do it now, it'll never happen.. 
I may need more info/context as to what each of these methods are for, but in my mind, when I see a method `do_x` and another method `do_xs`, I expect the latter to be something along the lines of `for xs.each { |x| do_x }`. ya know?

Comment (underscorgan) [inline lib/vanagon/component.rb]: so get_source is used for the primary source which ends up getting some special treatment as opposed to the secondary sources which are stored in get_sources.  The fact that they're being treated differently now doesn't mean they should be, but unraveling the differences that matter and ones that can be changed feels like it could be a huge yak.

For example, the primary source (get_source) is set via the project url, and all of the secondary are set via add_source.

I don't think we should block this work on figuring this out, but we can and should absolutely add a ticket in the VANAGON project.

Comment (mwaggett) [inline lib/vanagon/component.rb]: In that case, I think we should have less-confusing method names. Maybe `get_secondary_sources` instead of `get_sources`?

Comment (underscorgan) [inline lib/vanagon/component.rb]: Can that go in a separate PR after this one is merged? I tend to like to keep renaming/refactoring isolated from feature work so it's easier in the event things need to be reverted.

File: lib/vanagon/component.rb (hunk 4 above)
Comment (underscorgan) [inline lib/vanagon/component.rb]: would this be better?

```
# Which specific tools or command line invocations
# should be used to extract a given component's sources
```

File: lib/vanagon/component.rb (hunk 6 above)
Comment (mwaggett) [inline lib/vanagon/component.rb]: yeah. I didn't look closely enough and I think I thought `get_sources` was a new thing.

File: lib/vanagon/component.rb (hunk 4 above)
Comment (mwaggett) [inline lib/vanagon/component.rb]: what about
```
# The specific tools or command line invocations that
# should be used to extract a given component's sources
```
?

File: lib/vanagon/component.rb (hunk 6 above)
Comment (underscorgan) [inline lib/vanagon/component.rb]: ah fair

File: lib/vanagon/component.rb (hunk 4 above)
Comment (underscorgan) [inline lib/vanagon/component.rb]: 👍

Comment (underscorgan) [inline lib/vanagon/component.rb]: updated

File: lib/vanagon/component.rb [hunk 7]
```diff
@@ -215,7 +217,7 @@ def get_source(workdir) # rubocop:disable Metrics/AbcSize
         @dirname = './'
 
         # If there is no source, there is nothing to do to extract
-        @extract_with = ': no source, so nothing to extract'
+        extract_with << ': no source, so nothing to extract'
```
Comment (mckern) [inline lib/vanagon/component.rb]: I wonder if I could ditch this line entirely. An empty `Array` is a safe array, and it'd cut down on chatter and verbosity in the Makefile.

man_presses_fingers_together_thinking_something_cg4p41662c

Comment (underscorgan) [inline lib/vanagon/component.rb]: I think this could go away, but I think the makefile is still going to be a mess because of https://github.com/puppetlabs/vanagon/blob/ae1f319dbd48ed20f4e65929cf71b2d2208f67da/lib/vanagon/component/source/local.rb#L82

Comment (mckern) [inline lib/vanagon/component.rb]: FYI I'm not ditching this in this PR. It'd require strenuous testing due to how Makefiles are rendered (a lot of `&&`ing).

File: spec/lib/vanagon/component_spec.rb (hunk 3 above)
Comment (mckern) [inline spec/lib/vanagon/component_spec.rb]: Disregard these comments since decompression happens in the Makefile and there's no useful framework for testing Makefile rules. This is as good as I think it's gonna get, minus a nitpick or two.

**Comment by melissa** [comment]
when was this last tested?

**Comment by mckern** [comment]
About 10 minutes ago.
